Failure when queries involve filters with non-equatable types

We came across to an error upgrading trino to `444`.
One of our query that work on trino `443` now return an exception 
> java.lang.AssertionError: Type cannot be used in a comparison expression (should have been caught in analysis): Geometry

We managed to reproduce the bug 

 ```
create table iceberg.test.values1 as SELECT *
             FROM (VALUES (1, 1, 2),
                          (2, 2, 4),
                          (3, 22, 22),
                          (4, 45, 65),
                          (5, 6, 6)) AS a (id, lat, lon);

create table iceberg.test.values2 as SELECT *
             FROM (VALUES (1, 'value1'),
                          (2, 'value2'),
                          (3, 'value3'),
                          (4, 'value4'),
                          (5, 'value5')) AS a (id, name);

WITH geom
         AS (select st_geometryfromtext('POLYGON((0 0,0 10,10 10,10 0 ))') as geo),
     cte1 as (select id
           from iceberg.test.values1
           where ST_WITHIN(ST_POINT(lon, lat), (select geo from geom))),
     cte2 as (select *
           from iceberg.test.values2
           where id in (select id from cte1))
select *
from cte2;

```

the block above works fine on trino `443` but if run on trino `444` return the exception cited above

Here's a simpler reproduction:

```sql
WITH t(k) AS (SELECT 1 WHERE random() > 0)
SELECT *
FROM t
WHERE k IN (
	SELECT k
	FROM t
	WHERE cardinality((VALUES empty_approx_set())) = 0)
```

This is the error it produces:
```
java.lang.AssertionError: Type cannot be used in a comparison expression (should have been caught in analysis): HyperLogLog
	at io.trino.sql.planner.DomainTranslator$Visitor.createComparisonExtractionResult(DomainTranslator.java:676)
	at io.trino.sql.planner.DomainTranslator$Visitor.visitComparison(DomainTranslator.java:460)
	at io.trino.sql.planner.DomainTranslator$Visitor.visitComparison(DomainTranslator.java:301)
	at io.trino.sql.ir.Comparison.accept(Comparison.java:103)
	at io.trino.sql.ir.IrVisitor.process(IrVisitor.java:27)
	at io.trino.sql.planner.DomainTranslator$Visitor.lambda$visitLogical$0(DomainTranslator.java:343)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at io.trino.sql.planner.DomainTranslator$Visitor.visitLogical(DomainTranslator.java:344)
	at io.trino.sql.planner.DomainTranslator$Visitor.visitLogical(DomainTranslator.java:301)
	at io.trino.sql.ir.Logical.accept(Logical.java:69)
	at io.trino.sql.ir.IrVisitor.process(IrVisitor.java:27)
	at io.trino.sql.planner.DomainTranslator.getExtractionResult(DomainTranslator.java:298)
	at io.trino.sql.planner.EffectivePredicateExtractor$Visitor.visitFilter(EffectivePredicateExtractor.java:164)
	at io.trino.sql.planner.EffectivePredicateExtractor$Visitor.visitFilter(EffectivePredicateExtractor.java:120)
	at io.trino.sql.planner.plan.FilterNode.accept(FilterNode.java:74)
	at io.trino.sql.planner.EffectivePredicateExtractor$Visitor.visitProject(EffectivePredicateExtractor.java:205)
	at io.trino.sql.planner.EffectivePredicateExtractor$Visitor.visitProject(EffectivePredicateExtractor.java:120)
	at io.trino.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at io.trino.sql.planner.EffectivePredicateExtractor.extract(EffectivePredicateExtractor.java:117)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitFilteringSemiJoin(PredicatePushDown.java:1346)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitSemiJoin(PredicatePushDown.java:1286)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitSemiJoin(PredicatePushDown.java:146)
	at io.trino.sql.planner.plan.SemiJoinNode.accept(SemiJoinNode.java:141)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitFilter(PredicatePushDown.java:376)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitFilter(PredicatePushDown.java:146)
	at io.trino.sql.planner.plan.FilterNode.accept(FilterNode.java:74)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:423)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitProject(PredicatePushDown.java:276)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitProject(PredicatePushDown.java:146)
	at io.trino.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:423)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitPlan(PredicatePushDown.java:180)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitPlan(PredicatePushDown.java:146)
	at io.trino.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.trino.sql.planner.plan.OutputNode.accept(OutputNode.java:82)
	at io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:31)
	at io.trino.sql.planner.optimizations.PredicatePushDown.optimize(PredicatePushDown.java:140)
	at io.trino.sql.planner.optimizations.StatsRecordingPlanOptimizer.optimize(StatsRecordingPlanOptimizer.java:41)
	at io.trino.sql.planner.LogicalPlanner.runOptimizer(LogicalPlanner.java:301)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:264)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:236)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:231)
	at io.trino.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:498)
	at io.trino.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:478)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:416)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:272)
	at io.trino.dispatcher.LocalDispatchQuery.startExecution(LocalDispatchQuery.java:145)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$waitForMinimumWorkers$2(LocalDispatchQuery.java:129)
	at io.airlift.concurrent.MoreFutures.lambda$addSuccessCallback$12(MoreFutures.java:570)
	at io.airlift.concurrent.MoreFutures$3.onSuccess(MoreFutures.java:545)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1137)
	at io.trino.$gen.Trino_testversion____20240411_155039_75.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

The reason is that EffectivePredicateExtractor is deriving an incorrect predicate when a type that is not comparable or orderable is present. This is related to https://github.com/trinodb/trino/issues/21457 and https://github.com/trinodb/trino/pull/21494

--- a/core/trino-main/src/main/java/io/trino/sql/planner/EffectivePredicateExtractor.java
+++ b/core/trino-main/src/main/java/io/trino/sql/planner/EffectivePredicateExtractor.java
@@ -217,6 +217,7 @@ public class EffectivePredicateExtractor
                     .collect(toImmutableList());
 
             List<Expression> projectionEqualities = nonIdentityAssignments.stream()
+                    .filter(assignment -> assignment.getKey().type().isComparable() || assignment.getKey().type().isOrderable())
                     .filter(assignment -> Sets.intersection(SymbolsExtractor.extractUnique(assignment.getValue()), newlyAssignedSymbols).isEmpty())
                     .map(ENTRY_TO_EQUALITY)
                     .collect(toImmutableList());

--- a/core/trino-main/src/test/java/io/trino/sql/query/TestIssue21508.java
+++ b/core/trino-main/src/test/java/io/trino/sql/query/TestIssue21508.java
@@ -1,0 +1,39 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.query;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestIssue21508
+{
+    @Test
+    public void test()
+    {
+        try (QueryAssertions assertions = new QueryAssertions()) {
+            assertThat(assertions.query(
+                    """
+                    WITH t(k) AS (SELECT 1 WHERE random() >= 0)
+                    SELECT *
+                    FROM t
+                    WHERE k IN (
+                        SELECT k
+                        FROM t
+                        WHERE cardinality((VALUES empty_approx_set())) = 0)
+                    """))
+                    .matches("VALUES 1");
+        }
+    }
+}

